rllib/models/specs/specs_base.py

```diff
@@ -0,0 +1,212 @@
+
+        for expected_d, actual_d in zip(self._expected_shape, shape):
+            if isinstance(expected_d, int) and expected_d != actual_d:
+                raise ValueError(_INVALID_SHAPE.format(self._expected_shape, shape))
+
+        dtype = self.get_dtype(tensor)
+        if self.dtype and dtype != self.dtype:
+            raise ValueError(_INVALID_DTYPE.format(self.dtype, tensor.dtype))
+
+    @abc.abstractmethod
+    def get_shape(self, tensor: TensorType) -> Tuple[int]:
+        """Returns the shape of a tensor.
+
+        Args:
+            tensor: The tensor whose shape is to be returned.
+
+        Returns:
+            A `tuple` specifying the shape of the tensor.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_dtype(self, tensor: TensorType) -> Any:
+        """Returns the data type of a tensor.
+
+        Args:
+            tensor: The tensor whose data type is to be returned.
+
+        Returns:
+            The data type of the tensor.
+        """
+        raise NotImplementedError
+
+    @DeveloperAPI
+    def fill(self, fill_value: Union[float, int] = 0) -> TensorType:
+        """Creates a tensor filled with `fill_value` that matches the specs.
+
+        Args:
+            fill_value: The value to fill the tensor with.
+
+        Returns:
+            A tensor with the specified value that matches the specs.
+        """
+        full_shape = self._full_shape()
+        return self._full(full_shape, fill_value)
+
+    @abc.abstractmethod
+    def _full(self, shape: Tuple[int], fill_value: Union[float, int] = 0) -> TensorType:
+        """Creates a tensor with the given shape filled with `fill_value`. The tensor
+        dtype is inferred from `fill_value`.
+
+        Args:
+            shape: The shape of the tensor to be sampled.
+            fill_value: The value to fill the tensor with.
+
+        Returns:
+            A tensor with the specified value that matches the specs.
+        """
+        raise NotImplementedError
+
+    def _full_shape(self) -> Tuple[int]:
+        """Converts the expected shape to a shape by replacing the unknown dimension
+        sizes with a value of 1."""
+        sampled_shape = tuple()
+        for d in self._expected_shape:
+            if isinstance(d, int):
+                sampled_shape += (d,)
+            else:
+                sampled_shape += (1,)
+        return sampled_shape
+
+    def _parse_expected_shape(self, shape: str, shape_vals: Dict[str, int]) -> tuple:
+        """Converts the input shape to a tuple of integers and strings."""
+        expected_shape = tuple()
+
+        # check the validity of shape_vals and get a list of dimension names
+        d_names = shape.replace(" ", "").split(",")
+        self._validate_shape_vals(d_names, shape_vals)
+
+        for d in d_names:
+            d_value = shape_vals.get(d, None)
+            if d_value is None:
+                d_value = d
```

Comment:
So this is basically:
```
d_value = shape_vals.get(d, d)
```
and expected_shape can be a tuple of mixed integers and strings, right?
``how does expected_shape += d`` work?

Reply:
simplified the code. Thanks for pointing it out.